### PR TITLE
Various performance improvements

### PR DESCRIPTION
**Pull request by Theodlz**
In this PR, we address a number of performance issues:
- https://github.com/boom-astro/boom/issues/95 | deserializing avro alerts to structs for ZTF was much slower than expected. This is because ZTF does not use a schema registry like LSST, and instead does the good old "here's the schema in every single alert". Issue with that approach is that for every alert you receive, you spend time extracting the schema and the avro package likelt does a number of extra checks when it does so, which makes it quite slow. So instead here we add a "pseudo-schema registry", where the `ZtfAlertWorker` extract the schema from the first alert we process (similar to how we query the schema register for the first LSST alert we process) along with the index of the first byte of actual avro data. For all following alerts, we directly move to that index and use the cached schema. In case we get a failure when reading the alert, which could be caused by a schema change, we re-extract the schema and update the cache if we can read the alert with that new schema.
- https://github.com/boom-astro/boom/issues/96 | a lot of our structs have a ton of optional fields, that can be none. At the moment we serialize our structs to bson document and THEN sanitize them from all null fields. That is an expensive operation because we need to loop over all the key values of a document after converting to bson, create a new document that is mutable where we keep the valid fields. Instead, in this PR we simply use the serde_with crate to apply the `skip_serializing_if` attribute to all optional fields of our structs, which means that when they are serialized to bson the null fields are skipped to begin with, and we don't need to sanitize the bson.
- also we make some minor changes to the `spatial::xmatch` method to avoid a lot of unnecessary cloning and unwraps.
- there is also some cleanup in terms of package imports in the various files found in the `alert` dir.

TODOs:
- add some validation in the alert processing test, to check that after retrieving the documents from the DB we can't see some fields that were `None` in the avro. This is to check that the serde attribute is working as expected.

*PS: Will point to main once other PRs that this one depends on have been reviewed, right now I'm forced to point to another branch to not show changes that are irrelevant to this PR*

### CODE REVIEW

File: src/utils/db.rs [hunk 1]
```diff
@@ -26,21 +26,11 @@ pub async fn create_index(
...
-            }
-        }
-        cleaned_doc.insert(key, value);
-    }
-    cleaned_doc
+    // we removed all the sanitizing logic
+    // in favor of using serde's attributes to clean up the data
+    // ahead of time.
+    // TODO: drop this function entirely and avoid unwrapping
+    to_document(value).unwrap()
```
[the remark this answers is not in the document]
Comment (Theodlz) [reply src/utils/db.rs:33]: wouldn't work to just remove the unwrap, the function would then need to return a result

File: src/alert/ztf.rs [hunk 2]
```diff
@@ -8,13 +8,57 @@ use crate::{
...
+    let reader = Reader::new(cursor).map_err(SchemaRegistryError::InvalidSchema)?;
+    let schema = reader.writer_schema();
+
+    // Then, we look for the index of the start of the data
+    let mut cursor = std::io::Cursor::new(avro_bytes);
+
+    let mut buf = [0; 4];
+    cursor.read_exact(&mut buf).unwrap();
+    if buf != [b'O', b'b', b'j', 1u8] {
+        panic!("Magic bytes error");
```
[the remark this answers is not in the document]
Comment (Theodlz) [reply src/alert/ztf.rs]: Well this time it even suggested using an error type that didn't exist lmao

Still some work to do GitHub 😉

File: src/alert/ztf.rs [hunk 3]
```diff
@@ -347,19 +402,51 @@ pub struct ZtfAlertWorker {
...
 impl ZtfAlertWorker {
     pub async fn alert_from_avro_bytes(
         self: &mut Self,
         avro_bytes: &[u8],
     ) -> Result<ZtfAlert, AlertError> {
-        let reader = Reader::new(&avro_bytes[..]).map_err(AlertError::DecodeError)?;
+        // if the schema is not cached, get it from the avro_bytes
+        let (schema_ref, start_idx) = match self.cached_schema {
+            Some(ref schema) => (schema, self.cached_start_idx.unwrap()),
+            None => {
```
[the remark this answers is not in the document]
Comment (Theodlz) [reply src/alert/ztf.rs]: ahhhh this one is good, finally :)

File: src/utils/db.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Theodlz) [reply src/utils/db.rs:33]: still the same bad suggestion lol, just paraphrasing a comment and generating bad code isn't very useful 😅

File: config.default.yaml [hunk 4]
```diff
@@ -43,85 +42,38 @@ crossmatch:
...
       use_distance: false
       projection:
         _id: 1
-        coordinates.radec_str: 1
         parallax: 1
         parallax_error: 1
         phot_g_mean_mag: 1
         phot_bp_mean_mag: 1
         phot_rp_mean_mag: 1
-    - catalog: SDSS-V_master_20240719
```
Comment (jesaerys) [inline config.default.yaml:57]: Are all the catalogs removed here ones we don't need anymore?

**Review by jesaerys — APPROVED** [review]
I really like the approach you took to address #95 👍

Overall, really good insight on what was slowing things down.

File: config.default.yaml (hunk 4 above)
Comment (Theodlz) [reply config.default.yaml:57]: Basically when I was adding the crossmatch config to the codebase, I added some random catalogs I had on my machine. But, these aren't so relevant to what we do so I remove them here.
